### PR TITLE
Add all missing documentation

### PR DESCRIPTION
**Pull request by javierdelapuente**
Spec ISD124.

### Overview

This PR adds documentation for the NetBox charm:
 - charm/docs. Initial version following Diátaxis framework, that includes tutorials, how-tos, references and explanations.
 - charm/src-docs.

Extra: Added ca-certificates to rock as testing with github as a data source was failing.

Extra: Temporarily using operator-workflows@test-netbox as discourse-gatekeeper and operator-workflows have to be prepared to work with `paas-app-charmer` directory structure.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Manging Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: charm/README.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+<!--
+Avoid using this README file for information that is maintained or published elsewhere, e.g.:
+
+* metadata.yaml > published on Charmhub
+* documentation > published on (or linked to from) Charmhub
+* detailed contribution guide > documentation or CONTRIBUTING.md
+
+Use links instead.
+-->
```
Comment (mthaddon) [inline charm/README.md:15]: I think we can remove this commented section

File: charm/README.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+the freedom to deploy on the Kubernetes platform of their choice.
+
+For DevOps or SRE teams this charm will make operating NetBox simple
+and straightforward through Juju's clean interface. It will allow easy
+deployment into multiple environments for testing of changes.
+
+
+## Other resources
+
+<!-- If your charm is documented somewhere else other than Charmhub, provide a link separately. -->
```
Comment (mthaddon) [inline charm/README.md:44]: I think we can remove this comment

File: charm/docs/explanation/charm-architecture.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# NetBox charm architecture
+
+The NetBox charm initial version has been generated with the 
+help of the 12 Factor Project. The 12 Factor Project provides many
```
Comment (mthaddon) [inline charm/docs/explanation/charm-architecture.md:4]: Per https://docs.google.com/document/d/1uqm1S_ZRVCWwklHerfrIE-yrZb7bLLzjYsaepMM1pTQ/edit this is now named the paas-app-charmer project. We should check exactly how to refer to it in this context.

File: charm/docs/explanation/charm-architecture.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+The NetBox charm initial version has been generated with the 
+help of the 12 Factor Project. The 12 Factor Project provides many 
+of the functionalities needed by this charm like:
+- PostgreSQL integration
+- Django migrations
+- Ingress integration
+- COS (Prometheus metrics and Loki logs for gunicorn).
+
+For the static assets, gunicorn is used with the help of the  WhiteNoiseMiddleware.
```
Comment (mthaddon) [inline charm/docs/explanation/charm-architecture.md:11]: You've got two spaces between "the" and "WhiteNoiseMiddleware".

File: charm/docs/explanation/charm-architecture.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+There is only one container for each unit of NetBox, that runs the next 
+services managed by Pebble:
+- django. Runs gunicorn.
+- cron. Cron service that runs management commands for housekeeping and syncdatasource.
+- statsd_exporter. To expose gunicorn metrics.
+
+
```
[the remark this answers is not in the document]
Comment (javierdelapuente) [reply charm/docs/explanation/charm-architecture.md:19]: No. Updated to one.

File: .github/workflows/test.yaml [hunk 6]
```diff
@@ -5,7 +5,7 @@ on:
 
 jobs:
   unit-tests:
-    uses: canonical/operator-workflows/.github/workflows/test.yaml@main
```
Comment (arturo-seijas) [inline .github/workflows/test.yaml:8]: Is this intentional?

File: charm/docs/explanation/charm-architecture.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
+# NetBox charm architecture
+
+The NetBox charm initial version has been generated with the 
+help of the paas-app-charmer project. The paas-app-charmer project provides many
+of the functionalities needed by this charm like:
+- PostgreSQL integration
+- Django migrations
+- Ingress integration
+- COS (Prometheus metrics and Loki logs for gunicorn).
```
Comment (arturo-seijas) [inline charm/docs/explanation/charm-architecture.md:9]: Doesn't it include a dashboard too?

File: charm/docs/explanation/charm-architecture.md [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+For the static assets, gunicorn is used with the help of the WhiteNoiseMiddleware.
```
Comment (arturo-seijas) [inline charm/docs/explanation/charm-architecture.md:11]: A link to the `WhiteNoiseMiddleware` docs would be nice here

File: charm/docs/explanation/charm-architecture.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+Besides the integrations provided directly by the paas-app-charmer toolchain, the next
+integrations are implemented in NetBox:
+- SAML integration.
+- S3 integration.
+
+Currently Redis is configured using environment variables. This will be updated
+to a Redis integration.
+
+The NetBox charm is designed for a fully high availability (HA) environment, and many
```
Comment (arturo-seijas) [inline charm/docs/explanation/charm-architecture.md:27]: "and many instances can be run for it" This sounds a bit weird to me

File: charm/docs/explanation/charm-architecture.md [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+instances can be run for it. See [Configure Scripts and Reports for HA](../how-to/configure-scripts-reports.md)
+for the requirements to use scripts and reports in this HA configuration.
```
Comment (arturo-seijas) [inline charm/docs/explanation/charm-architecture.md:29]: s/in this HA/in HA

File: charm/docs/how-to/configure-scripts-reports.md [hunk 11]
```diff
@@ -0,0 +1,29 @@
+# Configure scripts and reports for high availability
+
+The NetBox charm is designed with high availability (HA) in mind. As there is no support
+for a shared filesystem, this implies that reports and scripts should not be uploaded as file uploads,
+as they will not work correctly with multiple instances and there is risk of data loss.
+
+To use scripts and reports, it is necessary to configure a Data Source, this can be of type
```
Comment (arturo-seijas) [inline charm/docs/how-to/configure-scripts-reports.md:7]: s/Source, this/Source, which

File: charm/docs/how-to/configure-scripts-reports.md [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+The NetBox charm is designed with high availability (HA) in mind. As there is no support
+for a shared filesystem, this implies that reports and scripts should not be uploaded as file uploads,
+as they will not work correctly with multiple instances and there is risk of data loss.
+
+To use scripts and reports, it is necessary to configure a Data Source, this can be of type 
+"Git" or "Amazon S3". The git repository https://github.com/netbox-community/customizations.git is
+a good source of examples of both scripts and reports.
+
+For S3 data sources, if the S3 server used is a compatible S3 server, the juju configuration
```
Comment (arturo-seijas) [inline charm/docs/how-to/configure-scripts-reports.md:11]: s/a compatible S3 server/S3 compatible

File: charm/docs/how-to/configure-scripts-reports.md [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+option `aws_endpoint_url` should be set to the url of the S3 server, as NetBox by default only
+allows using Amazon S3. This option limit the S3 compatible servers to just one possible
```
Comment (arturo-seijas) [inline charm/docs/how-to/configure-scripts-reports.md:13]: s/limit/limits

File: charm/docs/how-to/configure-scripts-reports.md [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+allows using Amazon S3. This option limit the S3 compatible servers to just one possible 
+S3 server. For example, if you use minio in the url `https://minio.example.com`, you
+should run `juju config netbox aws_endpoint_url=https://minio.example.com`.
+
+Once a data source for "Git" or "Amazon S3" is configured (menu Operations -> Data Sources),
+it should be synced (blue button in the top right of the data source screen).
+
+Once a data source is created and synced, scripts and reports can be created from the 
+data source (menu Customization -> reports & scripts). For that, the file in the data
+source should be selected, and the checkbox "Auto sync enabled" should be clicked.
```
Comment (arturo-seijas) [inline charm/docs/how-to/configure-scripts-reports.md:22]: s/selected, and/selected and

File: charm/docs/how-to/configure-scripts-reports.md [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+option `aws_endpoint_url` should be set to the url of the S3 server, as NetBox by default only
+allows using Amazon S3. This option limit the S3 compatible servers to just one possible 
+S3 server. For example, if you use minio in the url `https://minio.example.com`, you
+should run `juju config netbox aws_endpoint_url=https://minio.example.com`.
+
+Once a data source for "Git" or "Amazon S3" is configured (menu Operations -> Data Sources),
+it should be synced (blue button in the top right of the data source screen).
+
+Once a data source is created and synced, scripts and reports can be created from the 
+data source (menu Customization -> reports & scripts). For that, the file in the data
```
Comment (arturo-seijas) [inline charm/docs/how-to/configure-scripts-reports.md:21]: For that, the file in the data
source should be selected, and the checkbox "Auto sync enabled" should be clicked. /
For that, select the file in the data
source and click the checkbox "Auto sync enabled". 

The simpler the grammar, the better :)

File: charm/docs/tutorial/getting-started.md [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+## Deploy the NetBox charm
+
+Deploy the NetBox charm, with all its mandatory requirements (PostgreSQL, Redis and S3).
+
+### Deploy the charms:
+```
+juju deploy netbox
+```
+
+At this point netbox should be blocked as there it no S3 integration for
```
Comment (arturo-seijas) [inline charm/docs/tutorial/getting-started.md:41]: s/At this point netbox/At this point, netbox

File: charm/docs/tutorial/getting-started.md [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+storage, Redis nor PostgreSQL.
```
Comment (arturo-seijas) [inline charm/docs/tutorial/getting-started.md:42]: s/nor/or?

File: charm/docs/explanation/charm-architecture.md (hunk 7 above)
Comment (javierdelapuente) [reply charm/docs/explanation/charm-architecture.md:9]: Not yet, we are still discussing how to add specific metrics/log files with `paas-app-charmer`, and at that point at a proper dashboard (otherwise, it would be better is `paas-app-charmer` provides it).

File: .github/workflows/test.yaml (hunk 6 above)
Comment (javierdelapuente) [reply .github/workflows/test.yaml:8]: Kind of... as discourse-gatekeeper will be updated, it is necessary for NetBox to work for now until all is merged.

File: charm/docs/explanation/charm-architecture.md (hunk 7 above)
Comment (arturo-seijas) [reply charm/docs/explanation/charm-architecture.md:9]: Gotcha!

File: .github/workflows/test.yaml (hunk 6 above)
Comment (arturo-seijas) [reply .github/workflows/test.yaml:8]: Understood

File: charm/docs/explanation/charm-architecture.md (hunk 9 above)
Comment (arturo-seijas) [reply charm/docs/explanation/charm-architecture.md]: Yes. Thanks :)

File: charm/docs/how-to/configure-scripts-reports.md (hunk 15 above)
Comment (javierdelapuente) [reply charm/docs/how-to/configure-scripts-reports.md:21]: Thanks, it looks less convoluted!

**Review by arturo-seijas — DISMISSED** [review]

**Review by arturo-seijas — DISMISSED** [review]

File: charm/docs/explanation/charm-architecture.md [hunk 18]
```diff
@@ -0,0 +1,31 @@
+# NetBox charm architecture
+
+The NetBox charm initial version has been generated with the 
+help of the paas-app-charmer project. The paas-app-charmer project provides many
```
Comment (merkata) [inline charm/docs/explanation/charm-architecture.md:4]: Maybe like to the [paas-app-charmer piwheels](https://www.piwheels.org/project/paas-app-charmer/) or the [github page](https://github.com/canonical/paas-app-charmer/)?

File: charm/docs/explanation/charm-architecture.md [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+of the functionalities needed by this charm like:
+- PostgreSQL integration
+- Django migrations
+- Ingress integration
+- COS (Prometheus metrics and Loki logs for gunicorn).
+
+For the static assets, gunicorn is used with the help of the [WhiteNoiseMiddleware](https://whitenoise.readthedocs.io/en/stable/index.html).
+
+There is only one container for each unit of NetBox, that runs the following
```
Comment (merkata) [inline charm/docs/explanation/charm-architecture.md:13]: I don't think you need a comma here.

File: charm/docs/explanation/charm-architecture.md [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+services managed by Pebble:
+- django. Runs gunicorn.
+- cron. Cron service that runs management commands for housekeeping and syncdatasource.
```
Comment (merkata) [inline charm/docs/explanation/charm-architecture.md:16]: Is there a white space in `syncdatasource`?

File: charm/docs/how-to/contribute.md [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+# Enable DEBUG logging
+juju model-config logging-config="<root>=INFO;unit=DEBUG"
+# Deploy the charm (assuming you're on amd64)
+juju deploy ./charm/netbox_ubuntu-22.04-amd64.charm \
+  --resource django-app-image=localhost:32000/netbox:latest
+```
+
+### Configure NetBox
+
+NetBox required a Redis server to work correctly. This can
```
Comment (merkata) [inline charm/docs/how-to/contribute.md:98]: s/required/requires/

File: charm/docs/tutorial/getting-started.md [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+NetBox requires PostgreSQL to work. Deploy and integrate with:
+```
+juju deploy postgresql-k8s --channel 14/stable --trust
+juju integrate postgresql-k8s netbox
+```
+
+### Deploy s3-integrator
+
+NetBox requires a S3 integration for the uploaded files. This is because
+the NetBox charm is designed to work in a high availability (HA) configuration.
```
Comment (merkata) [inline charm/docs/tutorial/getting-started.md:82]: Maybe add a note how S3 enables HA, I would assume it is about storing things only.

File: rockcraft.yaml [hunk 23]
```diff
@@ -116,6 +116,8 @@ parts:
             chown -R 584792 django/app/reports
             chown -R 584792 django/app/scripts
     django-framework/dependencies:
+        overlay-packages:
+            - ca-certificates
```
Comment (merkata) [inline rockcraft.yaml:120]: Is this change intentional? It is not a doc change.

**Review by merkata — APPROVED** [review]
Good stuff, left minor remarks.

File: charm/docs/explanation/charm-architecture.md (hunk 18 above)
Comment (javierdelapuente) [reply charm/docs/explanation/charm-architecture.md:4]: Added link to the github page. Thanks!

File: charm/docs/explanation/charm-architecture.md (hunk 20 above)
Comment (javierdelapuente) [reply charm/docs/explanation/charm-architecture.md:16]: syncdatasource is the name of the django management command (see https://github.com/netbox-community/netbox/blob/develop/netbox/core/management/commands/syncdatasource.py)

File: charm/docs/tutorial/getting-started.md (hunk 22 above)
Comment (javierdelapuente) [reply charm/docs/tutorial/getting-started.md:82]: Added extra info about the reason for S3 (basically upload images to S3 instead to the local filesystem).

File: rockcraft.yaml (hunk 23 above)
Comment (javierdelapuente) [reply rockcraft.yaml:120]: I found the issue while testing git data sources (following the docs in this PR). I added it here for convenience (it is reflected in the PR info)

**Review by arturo-seijas — APPROVED** [review]

**Review by merkata — APPROVED** [review]
